### PR TITLE
add support for the updateNoteFields api

### PR DESCRIPTION
**Pull request by Twinov**
### Description
Adds support for the [updateNoteFields API](https://github.com/FooSoft/anki-connect#updatenotefields) to match the desktop AnkiConnect. There's a couple additional helper functions added as well that were needed to facilitate updateNoteFields.
### Example Request
```js
{
  action: 'updateNoteFields',
  version: 6,
  params: {
    note: {
      id: 1679779279219,
      fields: {
        Source: 'anime',
      },
      picture: [
        {
          data: image as base64 string,
          filename: 'screenshot.jpg',
          fields: ['Source'],
        },
      ],
    },
  },
}
```
### Limitations
Doesn't support URLs for files nor skipHash.
I only tested using images, though I uploaded them as audio/video to test the codepath and it worked fine.

### CODE REVIEW

File: app/src/main/java/com/kamwithk/ankiconnectandroid/ankidroid_api/IntegratedAPI.java
Comment (Aquafina-water-bottle) [inline app/src/main/java/com/kamwithk/ankiconnectandroid/ankidroid_api/IntegratedAPI.java:84]: It seems like `stored_files` only exists to check for duplicate files. I don't think Anki-Connect handles duplicate files unless specified with an additional [`deleteExisting`](https://github.com/FooSoft/anki-connect/blob/bf5c73b9cdec95337dc356b0d006b7847458c086/plugin/__init__.py#L673) field, so this logic can be safely removed.

**Review by Aquafina-water-bottle — CHANGES_REQUESTED** [review]
Hey @Twinov, thanks for the PR! Sorry it took a little while, but I left a few comments for the PR.

Additionally, I was talking with @KamWithK yesterday, and due to this PR and #28, we decided that it would be a good idea to provide something like a `docs/api.md`, where people can document what API calls exist, and any edge cases / unhandled parameters / etc. for each api call. Originally, the scope of this project was simply to provide an interface to Yomichan, but as indicated by these PRs. some developers are actually using this now.

There will be a PR soon to add that document, and once that happens, we'd like if you added `updateNoteFields` to that document before getting this properly merged. For example, one edge case that I noticed is that `deleteExisting` isn't handled, so that should be documented.

Nonetheless, this and #28 will be blocked until we get the skeleton of docs/api.md out (I'll aim to make the PR in a day or two).

File: app/src/main/java/com/kamwithk/ankiconnectandroid/ankidroid_api/IntegratedAPI.java
Comment (Twinov) [reply app/src/main/java/com/kamwithk/ankiconnectandroid/ankidroid_api/IntegratedAPI.java:84]: Actually the map is there because `mediaAPI.storeMediaFile()` doesn't produce the same filename as what's fed into it (at least when I was testing I think), so I cached the filename from the first time `mediaAPI.storeMediaFile()` was called. Otherwise, you'd store the same file multiple times within the same `updateNoteFields()` call under different filenames. More elegant solution would be to find out why `mediaAPI.storeMediaFile()` doesn't store with the exact filename it's called with, but that's why this map exists at least.

File: app/src/main/java/com/kamwithk/ankiconnectandroid/request_parsers/Parser.java
Comment (Twinov) [reply app/src/main/java/com/kamwithk/ankiconnectandroid/request_parsers/Parser.java:63]: What do you think about:
In Parser.java, just have a very dumb method `ArrayList<JsonObject> getUpdateNoteFieldsMedia(JsonObject raw_data)` where the JsonObjects are JsonArrays, and then all the work that's currently done in `getUpdateNoteFilesToAdd()` and `getUpdateNoteMediaFieldsToFilenames()` can be moved into `IntegratedAPI.updateNoteFields()`?

A more sustainable solution IMO would be to make a small class that encodes the audio/video/picture parameter from [here](https://github.com/FooSoft/anki-connect#addnote) and then have the parser parse into that, but tbh I don't have much Java experience and it'd be a lot of work if I'm the only one who's going to be using it, maybe consider it if someone wants to add media support to addNote though.

**Comment by Twinov** [comment]
Thanks for the review, your comments are good and I'll commit fixes soon, just want to know if you agree with my parser method suggestion before implementing. 
And yep can add docs once that's in master too.

File: app/src/main/java/com/kamwithk/ankiconnectandroid/request_parsers/Parser.java
Comment (Aquafina-water-bottle) [reply app/src/main/java/com/kamwithk/ankiconnectandroid/request_parsers/Parser.java:63]: I think this is fine as long as you mention in a comment that there could be work for improvement. If the storing media files can be moved to the function as mentioned earlier (lines 90-102), then getting the parser to parse the individual components shouldn't be difficult to do anyways.

File: app/src/main/java/com/kamwithk/ankiconnectandroid/ankidroid_api/IntegratedAPI.java
Comment (Aquafina-water-bottle) [reply app/src/main/java/com/kamwithk/ankiconnectandroid/ankidroid_api/IntegratedAPI.java:84]: Oh interesting, thanks for the clarification! In this case, would it be possible to put a comment explaining exactly that (and when the time comes, add that to the documentation too)?

Comment (Twinov) [reply app/src/main/java/com/kamwithk/ankiconnectandroid/ankidroid_api/IntegratedAPI.java:84]: Yeah can do.

**Comment by KamWithK** [comment]
Just a small reminder @Twinov that the docs pr was recently merged
So once you've made any necessary updates and added docs we can take another look at this PR 😃

**Comment by Twinov** [comment]
Made some pretty substantial changes to the updateNoteFields implementation that I think addresses the comments, but definitely needs to be rechecked. I did end up creating a small RequestMedia class to store each media object passed in when parsed.
Will check out the docs format and add soon. (as part of this PR)

File: app/src/main/java/com/kamwithk/ankiconnectandroid/ankidroid_api/IntegratedAPI.java
Comment (Twinov) [reply app/src/main/java/com/kamwithk/ankiconnectandroid/ankidroid_api/IntegratedAPI.java]: Removed it in the reworked implementation.

File: app/src/main/java/com/kamwithk/ankiconnectandroid/request_parsers/Parser.java
Comment (Twinov) [reply app/src/main/java/com/kamwithk/ankiconnectandroid/request_parsers/Parser.java]: Ended up creating a RequestMedia class for the parser to parse each audio/video/picture into, since it was easier than sending around the metadata to be honest. Let me know what you think.

File: app/src/main/java/com/kamwithk/ankiconnectandroid/ankidroid_api/IntegratedAPI.java
Comment (Twinov) [reply app/src/main/java/com/kamwithk/ankiconnectandroid/ankidroid_api/IntegratedAPI.java]: New implementation doesn't have the code duplication anymore by changing where the enclosed filename is created, so didn't make a helper function.

**Review by Aquafina-water-bottle — COMMENTED** [review]
Appreciate the changes! Everything overall looks fine now, but I do still have a question with how the `stored` variable is used, as commented above.

File: app/src/main/java/com/kamwithk/ankiconnectandroid/ankidroid_api/IntegratedAPI.java
Comment (Twinov) [reply app/src/main/java/com/kamwithk/ankiconnectandroid/ankidroid_api/IntegratedAPI.java]: You're right, everything in RequestMedia is final now and no need to keep metadata about whether file has been stored before.

**Comment by Twinov** [comment]
Rebased and resolved conflicts (thus the force push), addressed comments, and added docs. Thanks for the review, good comments.

File: app/src/main/java/com/kamwithk/ankiconnectandroid/ankidroid_api/MediaAPI.java
Comment (KamWithK) [inline app/src/main/java/com/kamwithk/ankiconnectandroid/ankidroid_api/MediaAPI.java:38]: Is this number maybe just a unique identifier?

Comment (Twinov) [reply app/src/main/java/com/kamwithk/ankiconnectandroid/ankidroid_api/MediaAPI.java:38]: Yes I think it's just a random unique ID, issue is that it's not accurate to how desktop Ankiconnect handles saving files. Desktop version saves files as the passed in filename.

**Comment by Aquafina-water-bottle** [comment]
Sorry again for the delay! Just looked through everything again, and it looks good to me! Thanks for responding to all of my earlier nitpicks :tada:

**Pull request merged**
